### ISSUE TITLE
Support non-registry mount sources

### ISSUE THREAD

**Issue by ktock** [containerd/stargz-snapshotter#668]
Stargz snapshotter should support more mount sources other than registry.
They include the following (discussed at https://github.com/moby/buildkit/pull/2685#issuecomment-1058888215).

- local OCI directory
- BuildKit Github Actions cache

**Comment by qi0523** [comment]
I want to implement P2P lazy pulling with containerd and stargz-snapshotter. Each node deploys a `agent` listening to 5001, which implements p2p distribution. 
When i use `ctr-remote image rpull 192.168.56.101:5001/qi0523/ubuntu:latest` to pull image lazily. I find that `index.json`, `manifest.json`, `config.json` can be downloaded from `agent`.

**Comment by ktock** [comment]
@qi0523 Could you provide a reproducible step and the log for that issue?
There are several docs for IPFS + stargz-snapshotter. Have you followed some of these steps?

- https://github.com/containerd/stargz-snapshotter/blob/v0.12.1/docs/ipfs.md
- https://github.com/containerd/nerdctl/tree/v1.0.0/examples/nerdctl-ipfs-registry-kubernetes/ipfs-stargz-snapshotter

**Comment by qi0523** [comment]
> @qi0523 Could you provide a reproducible step and the log for that issue? There are several docs for IPFS + stargz-snapshotter. Have you followed some of these steps?
> 
> * https://github.com/containerd/stargz-snapshotter/blob/v0.12.1/docs/ipfs.md
> * https://github.com/containerd/nerdctl/tree/v1.0.0/examples/nerdctl-ipfs-registry-kubernetes/ipfs-stargz-snapshotter

Thanks.

The `agent` is a simple component implemented by myself. 
It uses the image data under the directory`(/var/lib/containerd/io.containerd.content.v1.content/blobs/sha256)` to serve HEAD and GET requests, such as `http://xxx/v2/xxx/manifests/xxx`  and `http://xxx/v2/xxx/blobs/xxx`

**Comment by ktock** [comment]
@qi0523 So you're implementing your own registry API, right? Does that API fully implements OCI Distribution Spec? For successful lazy pulling, it needs to implement HTTP Range Request for blobs API at least.

Could you provide the log printed by stargz snapshotter (containerd-stargz-grpc) so that we can deeply investigate the issue?

You can also find the registry client logic at https://github.com/containerd/stargz-snapshotter/blob/v0.12.1/fs/remote/resolver.go

**Comment by qi0523** [comment]
> @qi0523 So you're implementing your own registry API, right? Does that API fully implements OCI Distribution Spec? For successful lazy pulling, it needs to implement HTTP Range Request for blobs API at least.
> 
> Could you provide the log printed by stargz snapshotter (containerd-stargz-grpc) so that we can deeply investigate the issue?
> 
> You can also find the registry client logic at https://github.com/containerd/stargz-snapshotter/blob/v0.12.1/fs/remote/resolver.go

Yes, i read the code of `distribution/distribution` and use `http.ServeContent(w, r, dgst.String(), time.Time{}, br)` to implement HTTP Range Request for blobs API.

![logs](https://user-images.githubusercontent.com/42767692/197721567-61f182bc-fa7f-4c7d-b8e6-1c85a958daa9.png)


The logs show that stargz snapshotter can not extract `sha256:73d944bd82`

**Comment by ktock** [comment]
@qi0523 The error logs look truncated. Could you copy them and post here as text?

**Comment by ktock** [comment]
@qi0523

> http: server gave HTTP response to HTTPS client

The error message says the `server gave HTTP response to HTTPS client`. So could you configure stargz snapshotter (/etc/containerd-stargz-grpc/config.toml) to use plain HTTP for connecting to your registry?

https://github.com/containerd/stargz-snapshotter/blob/v0.12.1/docs/overview.md#registry-mirrors-and-insecure-connection

**Comment by qi0523** [comment]
> @qi0523
> 
> > http: server gave HTTP response to HTTPS client
> 
> The error message says the `server gave HTTP response to HTTPS client`. So could you configure stargz snapshotter (/etc/containerd-stargz-grpc/config.toml) to use plain HTTP for connecting to your registry?
> 
> https://github.com/containerd/stargz-snapshotter/blob/v0.12.1/docs/overview.md#registry-mirrors-and-insecure-connection

Thanks.

Another question is that when we use stargz-snaphotter, `layer.tar` doesn't exist in directory `/var/lib/containerd/io.containerd.content.v1.content/blobs/sha256`. 

The data of `/var/lib/containerd-stargz-grpc/` can be used to serve blob GET request?

**Comment by ktock** [comment]
@qi0523 

> Another question is that when we use stargz-snaphotter, layer.tar doesn't exist in directory /var/lib/containerd/io.containerd.content.v1.content/blobs/sha256.

This is because of lazy pulling. containerd doesn't pull layer blobs but stargz-snapshotter lazily pulls chunks of layer blobs.

> The data of /var/lib/containerd-stargz-grpc/ can be used to serve blob GET request?

Yes. Blobs are stored under that directory and they are chunked. Note that some chunks of blobs can be lack until stargz-snapshotter completes downloading the entire blob contents to the node.

The followings are some of packages managing them:

- https://github.com/containerd/stargz-snapshotter/tree/v0.12.1/cache
- https://github.com/containerd/stargz-snapshotter/tree/v0.12.1/fs/reader
- https://github.com/containerd/stargz-snapshotter/blob/v0.12.1/fs/layer/layer.go

**Comment by qi0523** [comment]
> @qi0523
> 
> > Another question is that when we use stargz-snaphotter, layer.tar doesn't exist in directory /var/lib/containerd/io.containerd.content.v1.content/blobs/sha256.
> 
> This is because of lazy pulling. containerd doesn't pull layer blobs but stargz-snapshotter lazily pulls chunks of layer blobs.
> 
> > The data of /var/lib/containerd-stargz-grpc/ can be used to serve blob GET request?
> 
> Yes. Blobs are stored under that directory and they are chunked. Note that some chunks of blobs can be lack until stargz-snapshotter completes downloading the entire blob contents to the node.
> 
> The followings are some of packages managing them:
> 
> * https://github.com/containerd/stargz-snapshotter/tree/v0.12.1/cache
> * https://github.com/containerd/stargz-snapshotter/tree/v0.12.1/fs/reader
> * https://github.com/containerd/stargz-snapshotter/blob/v0.12.1/fs/layer/layer.go

Hi, i try to use the data of `/var/lib/containerd-stargz-grpc/` to serve blob GET request.  
Generally, the URL of GET request is `http://xxxxx/blobs/sha256:xxx`. So, we can only know the digest of the blob. 
Although the above packages can manage the stargz layer, the mapping of blobs‘ digest and stargz layer is not maintained. How can i find out stargz layer by using the digest of the blob(`sha256:xxx`)?
